hardware_interface/include/hardware_interface/actuator_interface.hpp

```diff
@@ -115,33 +120,202 @@ class ActuatorInterface : public rclcpp_lifecycle::node_interfaces::LifecycleNod
    * \returns CallbackReturn::SUCCESS if required data are provided and can be parsed.
    * \returns CallbackReturn::ERROR if any error happens or data are missing.
    */
-  virtual CallbackReturn on_init(const HardwareInfo & /*hardware_info*/)
+  virtual CallbackReturn on_init(const HardwareInfo & hardware_info)
   {
+    info_ = hardware_info;
+    import_state_interface_descriptions(info_);
+    import_command_interface_descriptions(info_);
     return CallbackReturn::SUCCESS;
   };
 
+  /**
+   * Import the InterfaceDescription for the StateInterfaces from the HardwareInfo.
+   * Separate them into the possible types: Joint and store them.
+   */
+  virtual void import_state_interface_descriptions(const HardwareInfo & hardware_info)
+  {
+    auto joint_state_interface_descriptions =
+      parse_state_interface_descriptions(hardware_info.joints);
+    for (const auto & description : joint_state_interface_descriptions)
+    {
+      joint_state_interfaces_.insert(std::make_pair(description.get_name(), description));
+    }
+  }
+
+  /**
+   * Import the InterfaceDescription for the CommandInterfaces from the HardwareInfo.
+   * Separate them into the possible types: Joint and store them.
+   */
+  virtual void import_command_interface_descriptions(const HardwareInfo & hardware_info)
+  {
+    auto joint_command_interface_descriptions =
+      parse_command_interface_descriptions(hardware_info.joints);
+    for (const auto & description : joint_command_interface_descriptions)
+    {
+      joint_command_interfaces_.insert(std::make_pair(description.get_name(), description));
+    }
+  }
+
```

Comment:
Well these methods are pretty much repeating in Sensor, System and Actuator, and if they want to override may be they can do it by overriding the `on_init` right?.

Anway, if repeating code is not an issue for you guys, fine for me :)

Reply:
My initial intend was to provide default functionality by the framework but give the user possibility to override the functions for exporting the stateinterfaces in case they want to do some hardware specific stuff or order them differently. I am not sure which way we want to go, i would be interested in your opinions on this.

Reply:
Personally, I would separate them into helper methods that would work for  `joints`, `gpois` etc, and reuse them across all the hardware interfaces, and if the user wants to override it, then he can do it in the derived class `on_init` method

This way you would be able to automatically export all types of interfaces.

Anyway, this is just an implementation detail. What do you think?